libraries/AP_FlashIface/AP_FlashIface_JEDEC.cpp

```diff
@@ -0,0 +1,849 @@
+    if (!detect_device()) {
+        hal.console->printf("Failed to detect flash device: %s\n", supported_devices[_dev_list_idx].name);
+        return false;
+    }
+
+    // Configuring Device involved setting chip to correct WSPI mode
+    // i.e. 4-4-4
+    if (!configure_device()) {
+        hal.console->printf("Failed to config flash device: %s\n", supported_devices[_dev_list_idx].name);
+        return false;
+    }
+
+    hal.console->printf("Detected Flash Device: %s\n", supported_devices[_dev_list_idx].name);
+    return true;
+}
+//////////////////////////////////////////////////////
+////////////////// Internal Methods //////////////////
+//////////////////////////////////////////////////////
+void AP_FlashIface_JEDEC::reset_device()
+{
+    // Get chip out of XIP mode
+    AP_HAL::QSPIDevice::CommandHeader cmd;
+    uint8_t buf[1];
+    _dev->get_semaphore()->take_blocking();
+
+    /* Attempting a reset of the XIP mode, it could be in an unexpected state
+     because a CPU reset does not reset the memory too.*/
+    /* Resetting XIP mode by reading one byte without XIP confirmation bit.*/
+    cmd.cmd   = 0U;
+    cmd.alt   = 0xFFU;
+    cmd.addr  = 0U;
+    cmd.dummy = 6U;
+    cmd.cfg   = AP_HAL::QSPI::CFG_CMD_MODE_NONE |
+                AP_HAL::QSPI::CFG_ADDR_SIZE_24 |
+                AP_HAL::QSPI::CFG_ADDR_MODE_FOUR_LINES |
+                AP_HAL::QSPI::CFG_DATA_MODE_FOUR_LINES |
+                AP_HAL::QSPI::CFG_ALT_MODE_FOUR_LINES |
+                AP_HAL::QSPI::CFG_ALT_SIZE_8;
+    _dev->set_cmd_header(cmd);
+    _dev->transfer(nullptr, 0, buf, 1);
+
+    /// NOTE: This is Vendor specific, we haven't read the parameter table
+    /// yet, so don't know how we can reset the chip.
+    // Various methods for Soft Reset are at Ref. JESD216 6.4.19
+    /* Quad line CMD_RESET_ENABLE command.*/
+    cmd.cmd   =  CMD_RESET_ENABLE;
+    cmd.cfg   =  AP_HAL::QSPI::CFG_CMD_MODE_FOUR_LINES;
+    cmd.addr  =  0;
+    cmd.alt   =  0;
+    cmd.dummy =  0;
+    _dev->set_cmd_header(cmd);
+    _dev->transfer(nullptr, 0, nullptr, 0);
+
+    /* Quad line CMD_RESET_MEMORY command.*/
+    cmd.cmd   =  CMD_RESET_MEMORY;
+    cmd.cfg   =  AP_HAL::QSPI::CFG_CMD_MODE_FOUR_LINES;
+    cmd.addr  =  0;
+    cmd.alt   =  0;
+    cmd.dummy =  0;
+    _dev->set_cmd_header(cmd);
+    _dev->transfer(nullptr, 0, nullptr, 0);
+
+    /* Single line CMD_RESET_MEMORY command.*/
+    cmd.cmd    =  CMD_RESET_ENABLE;
+    cmd.cfg    =  AP_HAL::QSPI::CFG_CMD_MODE_ONE_LINE;
+    cmd.addr   =  0;
+    cmd.alt    =  0;
+    cmd.dummy  =  0;
+    _dev->set_cmd_header(cmd);
+    _dev->transfer(nullptr, 0, nullptr, 0);
+
+
+    /* Single line N25Q_CMD_RESET_MEMORY command.*/
+    cmd.cmd   =  CMD_RESET_MEMORY;
+    cmd.cfg   =  AP_HAL::QSPI::CFG_CMD_MODE_ONE_LINE;
+    cmd.addr  =  0;
+    cmd.alt   =  0;
+    cmd.dummy =  0;
+    _dev->set_cmd_header(cmd);
+    _dev->transfer(nullptr, 0, nullptr, 0);
+
```

Comment:
At least on the winbond chips there is a prescribed delay between reset and doing anything else. You probably need some kind of delay in here